### ISSUE TITLE
Read latency can sometimes go up to 1 second

### ISSUE THREAD

**Issue by crazyzhou** [pravega/pravega#6440]
**Describe the bug**
When we start the local Pravega standalone, and perform `readNextEvent()` after `writeEvent()`, very rarely(under 0.1% in our test) it cannot return the event inside the 1000-millisecond timeout duration.
We have tested that there is no data loss because when call the `readNextEvent` again it will return the correct event. The real question is that why the read latency is so high up to over a second given the Pravega standalone is a pure in-memory system. Is there anything inside segment store that hangs the read operation?

**To Reproduce**
Start the standalone and run the following code, it will sometimes print `null` in the middle and probably will not read till `9999`.
```
    for (int i = 0; i < 10000; i++) {
        writeData(writer, String.valueOf(i));
//        writer.flush();
        System.out.println(readData(reader));
    }

    private static void writeData(EventStreamWriter<String> writer, String event) {
        final CompletableFuture<Void> writeFuture = writer.writeEvent(event);
        writeFuture.join();
    }

    private static String readData(EventStreamReader<String> reader) {
        EventRead<String> eventRead = reader.readNextEvent(1000);
        return eventRead.getEvent();
    }
```

**Screenshots**

**Additional information**

**Comment by tkaitchuck** [comment]
The absence of an event retrieved from a reader does not mean that no events exist, merely that one was not obtained before the provided timeout. To check for the absence of events a better way is to look at the ReaderGroupMetrics for the unreadBytes. If this is 0, then there is nothing to read. (Note that metrics have a delay).

Second, returning `null` for an event doesn't actually mean it timed out. Looking at the code there it isn't checking the `isCheckpoint` flag, which if set there will be a null event even if the timeout has not occured. 

If it is actually taking more than 1s to obtain an event in the middle of a stream, my first guess would be scaling. If a stream splits into multiple segments in response to a fast write rate, there is some reader group wide synchronization that needs to happen to coordinate all the readers which involves multiple round trips, which have a throttle on them so they don't actually go quickly. That being said, looking at the above code checkpoint seems more likely.

**Comment by crazyzhou** [comment]
Hi @tkaitchuck , I understand and also checked all these things.
First, you can see the sample app above, it is performing a read after a write, we also wait until the write future returns, so guarantees the data is persisted and there is event can be obtained.
Second, I also suspected the checkpoint thing, so in the reader group, I have explicitly disabled the auto checkpointing, but the result is still the same. I also checked again today that the `isCheckpoint` flag is false for empty returns.
Third, I also suspected the scaling differs, so I use the fixed one segment in all my tests.

**Comment by tkaitchuck** [comment]
Ok. That leaves two possibilities, ether the issue is on the server or the client. The metrics produced by the server should tell what it perceives the latency as, and can give a breakdown of where it is going if it is high. 
The other possibility, which seems suspicious given the number 1s is that there is a missed notification on the client and it occasionally isn't waking up properly in response to data from the server. This is easy to test:
Run the same setup but set the timeout to say 200millis and check if you ever observe 5 consecutive timeouts. If not, the client has a bug which is the issue rather than the server.

**Comment by tkaitchuck** [comment]
To clarify my investigation: May I assume there is only one reader in the reader group and it is using only one thread when you have observed this issue?

**Comment by crazyzhou** [comment]
Yes, one reader, one segment, single thread, writing small events. When I change to 200ms, there are 4 null/timeout returns and they are NOT consecutive.

**Comment by shafqatevo** [comment]
Hi @crazyzhou was there any conclusion for this issue?
